{% extends "base.html" %}
{% load static %} 

{% block sidebar %}
<p><span class="pageName">Definitions</span></p>

<p class="smalltext"><span class="subHeader">Genomic Islands</span> - large genomic regions (typically &gt; 8kb), that are thought to have horizontal origins. These regions can often contain genes that are related to antibiotic resistance or virulence.</p>

<p class="smalltext"><span class="subHeader">Sequence Composition GI Prediction Methods</span> - bacterial genomes tend to naturally have differences in sequence composition such as GC% and codon bias. Therefore, a region within a genome that has abnormal sequence composition could indicate that it originated from another genome. Various tools (incl. SIGI-HMM and IslandPath-DIMOB) look for regions with abnormal sequence composition.  </p>
		  
<p class="smalltext"><span class="subHeader">Comparative Genomic GI Prediction Methods</span> - genomic regions that are not present within related strains suggest that the region was horizontally transferred. IslandPick is a method that automatically selects comparison genomes that are within a proper evolutionary distance and identifies regions that are unique to only a single genome. </p>
		  
<p class="smalltext"> <span class="subHeader">Mobility Genes </span>- are genes related to the movement of mobile genetic elements (incl. genomic islands), such as transposases and integrases. </p>
		  
{% endblock sidebar %}

{% block content %}
<p class="bodyText"><span class="pageName">About</span> (<a href="{% url 'faq' %}">Visit our FAQs</a>)</p>

<p class="bodyText"><span class="subHeader">IslandViewer</span></p>

<p class="bodyText">This web site was developed so that researchers could easily  <a href="{% url 'browse' %}">view and download <em>genomic islands</em></a> for all published sequenced bacterial and archaeal genomes that have been predicted using the the currently most accurate GI prediction methods. Users can also <a href="{% url 'uploadform' %}">upload</a> their own unpublished genomes for analysis. The source code and entire GI data sets are available for <a href="{% url 'download' %}">download</a> and <a href="{% url 'acknowledgements' %}">acknowledgment</a> information is available for those who use our resources . Please <a href="{% url 'contactus' %}">contact us</a> with any questions or comments. Note, this is ISlandViewer 3, the older IslandViewer 2 is still available for any old analysis needs.</p>

<p class="subHeader">IslandPick<a name="islandpick" id="islandpick"></a></p>

<p class="bodyText">In a recent study (<a href="http://www.biomedcentral.com/1471-2105/9/329" target="_blank">Langille et al., 2008</a>), we describe a new method called <a href="{% url 'islandpick' %}">IslandPick</a> that uses a<em> comparative genomic GI prediction method </em>  to develop  stringent data sets of GIs and non-GIs. These positive and negative data sets were used to evaluate several <em>sequence composition GI prediction methods</em> and showed that SIGI-HMM and IslandPath-DIMOB were shown to have the highest overall accuracy. In addition, IslandPick had the most agreement with an independent data set of previously published genomes indicating that it was a highly precise method for GI prediction. IslandPick requires  several phylogentically related genomes to be sequenced to be able to make a prediction; therefore, predictions will not be available for many genomes. Non-default comparison genomes may be chosen using the <a href="{% url 'islandpick' %}">IslandPick link</a>. For more information please see <a href="http://www.biomedcentral.com/1471-2105/9/329" target="_blank">Langille et al., 2008</a> and/or our <a href="datasets/Islandpick_GI_dataset_overview.pdf" target="_blank">overview presentation</a>. </p>

<p class="subHeader">SIGI-HMM<a name="sigi" id="sigi"></a></p>

<p class="bodyText"><a href="http://www.tcs.informatik.uni-goettingen.de/colombo-sigihmm" target="_blank">SIGI-HMM</a> (see <a href="http://www.biomedcentral.com/1471-2105/7/142" target="_blank">Waack et al., 2006</a>) is a <em>sequence composition GI prediction method</em> that is part of the Columbo package. This method uses a Hidden Markov Model (HMM) and measures codon usage to identify possible GIs. In a recent study, SIGI-HMM was shown to have the highest precision and overall accuracy, out of six tested <em>sequence composition GI prediction methods</em> (<a href="http://www.biomedcentral.com/1471-2105/9/329" target="_blank">Langille et al., 2008</a>). </p>

<p class="subHeader">IslandPath-DIMOB<a name="dimob" id="dimob"></a></p>
		  
<p class="bodyText"><a href="http://www.pathogenomics.sfu.ca/islandpath/" target="_blank">IslandPath</a> (see <a href="http://bioinformatics.oxfordjournals.org/cgi/content/abstract/19/3/418" target="_blank">Hsiao et al., 2003</a>) was originally designed to aid to the identification of prokaryotic genomics islands (GIs), by visualizing several common characteristics of GIs such as abnormal sequence composition or the presence of genes that functionally related to mobile elements (termed <em>mobility genes</em>). </p>
		  		  
<p class="bodyText">Our subsequent studies (see <a href="http://www.biomedcentral.com/1471-2105/9/329" target="_blank">Langille et al., 2008</a>), showed that dinucleotide sequence composition bias   and the presence of <em>mobility genes </em>were    good indicators for identifying GIs. In fact, it was tied with SIGI-HMM for having the highest overall accuracy and traded a slightly lower precision for higher recall. </p>

<p class="subHeader">Curated Virulence Factors and Antibiotic Resistance Genes<a name="vir" id="vir"></a></p>
		  
<p class="bodyText">Curated virulence factor annotations are incorporated from the 2012 release of the <a href="http://www.mgc.ac.cn/VFs/">Virulence Factor Database, VFDB </a> (see <a href="http://nar.oxfordjournals.org/content/40/D1/D641.full">Chen et al., 2012</a>), <a href="http://www.patricbrc.org">PATRIC</a> (see <a href="http://nar.oxfordjournals.org/content/42/D1/D581.long">Wattam et al., 2014</a>) and <a href="http://www.phidias.us/victors/">Victors</a>. Antibiotic resistance gene classifications were downloaded from the <a href="http://arpcard.mcmaster.ca/">Comprehensive Antibiotic Resistance Database (CARD)</a> (see <a href="http://aac.asm.org/content/57/7/3348.long">McArthur et al., 2013</a>). These curated annotations are available for visualization on genome images within the scatter plot, as well as for download in the various downloadable files. Within the linear genome view, links are provided links to the respective entries from the annotation source for more information. In downloadable files, the source of each annotation is denoted using its acronym.</p>

<p class="subHeader">Homologs of Virulence Factors and Antibiotic Resistance Genes</p>

<p class="bodyText">Homologs of curated virulence factor genes and resistance genes have been provided for genomes missing such curated data. Resistance gene homologs were determined using the <a href="http://arpcard.mcmaster.ca/?q=CARD/tools/RGI">Resistance Gene Identifier</a> tool available through the CARD database. Virulence factor homologs were identified in close relatives of genomes with curated data based on a reciprocal best blast hit (RBBH) approach with very stringent cutoff values: e-value cutoff of 1e-10, >90% sequence similarity, and >80% coverage. All homologs of curated annotations are indicated within the scatter plot in lighter shades and are also available for download. In downloadable files, the source for these homologs may be indicated by "RGI" for resistance genes and "BLAST" for virulence factors. </p>

<p class="subHeader">Pathogen-Associated Genes<a name="pag" id="pag"></a></p>

<p class="bodyText">Pathogen-associated genes are those genes only found in pathogens and not in non-pathogens according to set criteria. A study conducted by Ho Sui et al. in 2009 compiled a list of these pathogen-associated genes (see <a href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0008094">Ho Sui et al., 2009</a> for more information and find data for <a href="http://www.pathogenomics.sfu.ca/pathogen-associated/">download here</a>). An update of this analysis was performed to include more recent genomes (all NCBI complete bacterial genomes from September 3, 2014). These annotations are also indicted within the scatter plot to compare against GI predictions. For robustness we included only those pathogen-associated genes found conserved across at least three genera (but never in non-pathogens of any genera). In downloadable files, the source for pathogen-associated genes may be indicated by the acronym "PAG".</p>
		  
<p class="bodyText"><span class="subHeader">Analysis Considerations<a name="considerations" id="considerations"></a> </span></p>

<p class="bodyText"><a href="{% url 'browse' %}">IslandViewer</a> integrates two <em>sequence composition GI prediction methods</em> SIGI-HMM and IslandPath-DIMOB, and a single <em>comparative GI prediction method</em> IslandPick. These methods have varying advantages and disadvantages. </p>
		  
<p class="bodyText"><em>Sequence composition GI prediction methods</em> may have difficulty detecting ancient GIs due to the amelioration of the foreign DNA to the host genome over time. Also, these methods will have difficulty detecting GIs that have originated from a genome with  similar sequence composition as the host genome. Lastly, these methods can make false predictions due to the normal variation in sequence composition that can occur in bacterial genomes.</p>

<p class="bodyText"><em>Comparative GI prediction methods  </em>depend heavily on the genomes that are chosen for comparison. For instance, the selection of very similar genomes will result in the prediction of only recently inserted GIs, while comparing with more distantly related genomes will detect many more GIs (including recent and ancient GIs), but may increase the chance of false prediction. <a href="{% url 'islandpick' %}">IslandPick</a> uses several different cutoffs to automatically select genomes for comparison, but users have the choice to select different comparison genomes based on their own insights.</p>

<p class="bodyText">In general, we recommend that users take advantage of the ability to view GI predictions from all three methods in a single integrated view. To aid this further, we have highlighted islands that are predicted by one or more tools as <span class="red">red</span> around the outer circle. </p>

<p class="subHeader">Conducting Further Analysis<a name="analysis" id="analysis"></a> </p>

<p class="bodyText">We strongly encourage researchers to conduct further analyses of any GIs displayed by IslandViewer. As outlined above, in <a href="{% url 'about' %}#considerations">Analysis Considerations</a>, there are many factors that can lead to false prediction   by any GI prediction program. In addition, any GIs that are very close to each other in the genome may actually be one large GI insertion (or vice versa) and the exact boundaries or insertion sites should be inspected in more detail. We recommend the use of the following computational tools to aid in further analysis of genomic islands.</p>

<p class="bodyText"><a href="http://www.sanger.ac.uk/Software/Artemis/">Artemis</a>, is a popular and well developed genome browser and annotation tool. IslandViewer allows users to download predicted GIs in a GenBank file format that can be opened directly in Artemis. The GIs will appear in Artemis  using  the same colour scheme used in IslandViewer (i.e. green for IslandPick, orange  for SIGI-HMM, etc.). An easy to view list of islands can be produced in Artemis by using the &quot;Feature Selector&quot; under the &quot;Select&quot; menu and filling in:</p>

		  <ul class="bodyText">
		    <li>Key = &quot;misc_feature&quot;</li>
	        <li> Qualifier = &quot;note&quot;</li>
		    <li>Containing this text = &quot;Genomic Island&quot;</li>
		    </ul>		  

<p class="bodyText">Along with the inspection of genes neighbouring and within GIs, sequence composition graphs, such as GC content and Karlin Signature Difference, can be added via the Graph menu. Users with their own GI predictions or other genome features, such as IS or direct repeat elements, can add their data to Artemis after constructing a fairly <a href="http://www.sanger.ac.uk/Software/Artemis/manual/concepts.html#CONCEPTS-TABFILE">simple input file</a>. </p>

<p class="bodyText"><a href="http://www.pathogenomics.sfu.ca/islandpath/">IslandPath</a>, displays many features that are often associated with genomic islands, such as various types of sequence composition bias, tRNAs, and integrases and transposases. These features are shown together in a single integrated view and may help in determining the location of GIs. IslandViewer contains a direct link to IslandPath for the genome being viewed on the left column of every results page. </p>

<p class="bodyText"><a href="http://asap.ahabs.wisc.edu/mauve/">Mauve</a>, is a whole genome alignment tool that can be used to view genome rearrangements and large insertions in related genomes. By using the IslandViewer GenBank download file along with your choice of other closely related bacteria genomes, GIs can be viewed in the context of other genomes and conserved regions surrounding these regions can be inspected further. 

                  <br />

<p class="subHeader">How to link to IslandViewer results<a name="analysis" id="analysis"></a> </p>

<p class="bodyText">Websites are welcome to link to island predictions in IslandViewer. The RefSeq accession can be used to link to the page in the format: http://www.pathogenomics.sfu.ca/islandviewer/results.php?query_input=NC_XXXXXX Note, that the accession version is optional and if not given the most recent version will be used (e.g NC_003997 is equivalent to NC_003997.3).</p>		  
{% endblock content %}
